docs: add explanation for providedIn any

santoshyadavdev opened the pull request:
Fixes #35179

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #35179


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

santoshyadav198613 commented:
Hi @gkalpak and @kapunahelewong ,

Do we need to do something different for images, the error is due to path of image.

kapunahelewong commented on aio/content/guide/providers.md:
```diff
@@ -52,6 +52,14 @@ Any component created within a lazy loaded module’s context, such as by router
 
 Though you can provide services by lazy loading modules, not all services can be lazy loaded. For instance, some modules only work in the root module, such as the Router. The Router works with the global location object in the browser.
 
+With Angular 9, it is possible to provide a new instance of service with each lazy loaded module. You can change the UserService code like:
```
```suggestion
With Angular version 9, you can provide a new instance of service with each lazy loaded module. The change to `UserService` would be as follows:
```

santoshyadav198613 replied:
Done.

kapunahelewong commented on aio/content/guide/providers.md:
```diff
@@ -52,6 +52,14 @@ Any component created within a lazy loaded module’s context, such as by router
+
+<code-example path="providers/src/app/user.service.2.ts"  header="src/app/user.service.ts"></code-example>
+
+Remember this is available for User defined services, and all eager loaded module will share an singleton insance, as shown in image below.
```
```suggestion
Using `providedIn: 'any'` is available for user defined services, or services.... . 
With this configuration, all eagerly loaded modules share a singleton instance, however lazy loaded modules each get their own unique instance, as in the following diagram.
```
Could you define user defined services?

santoshyadav198613 replied:
Done and removed this text to avoid confusion.

kapunahelewong reviewed (changes requested):
Thank you so much, @santoshyadav198613! Just a couple of suggestions on the writing. Regarding the image, could you use colors and styles like the ones in these images? 
https://angular.io/generated/images/guide/dependency-injection/injectors.svg
https://angular.io/generated/images/guide/inputs-outputs/input.svg
https://angular.io/generated/images/guide/inputs-outputs/input-diagram-target-source.svg
https://angular.io/generated/images/guide/inputs-outputs/input-output-diagram.svg

Those are all relatively recent additions where we're trying to use the Angular colors. If your software doesn't have what you need to do it, I'm happy to help you with it. Regarding the content of the diagram, could you also explain in the doc what's in each (like what's the `ConfigService`, etc.). If we have diagrams, we need to have a full explanations so that the content is still there for those using assistive technology like screen readers or those who can't see the images. That way the images are an enhancement (and not vital to understanding the concepts for people who can't refer to them).

kapunahelewong commented:
About the error loading the image, could it be that the extension of the file is all caps?

santoshyadav198613 commented:
> About the error loading the image, could it be that the extension of the file is all caps?

Thanks it worked now, changed to svg.

mary-poppins commented:
You can preview af7e1a1 at https://pr35283-af7e1a1.ngbuilds.io/.
You can preview 657a19c at https://pr35283-657a19c.ngbuilds.io/.

kapunahelewong commented on aio/content/guide/providers.md:
```diff
@@ -52,6 +52,14 @@ Any component created within a lazy loaded module’s context, such as by router
+
+<code-example path="providers/src/app/user.service.2.ts"  header="src/app/user.service.ts"></code-example>
+
+Using `providedIn: 'any'` is available for provider scope, With this configuration, all eagerly loaded modules share a singleton instance, however lazy loaded modules each get their own unique instance, as in the following diagram.
```
How about:

> With `providedIn: 'any'`, all eagerly loaded modules share a singleton instance, however lazy loaded modules each get their own unique instance, as in the following diagram.

santoshyadav198613 replied:
Made some changes to image and changes the text.

kapunahelewong reviewed (changes requested):
Just one suggestion for the copy. Also, nice work on the diagram. I've attached a screenshot with some small notes on the diagram. Thanks so much for all your hard work, @santoshyadav198613!!
<img width="646" alt="santosh-img-edits" src="https://user-images.githubusercontent.com/4116963/74273391-6342e300-4cde-11ea-80b1-ede2ef10f713.png">

mary-poppins commented:
You can preview de34c73 at https://pr35283-de34c73.ngbuilds.io/.

mary-poppins commented:
You can preview 5b03ac7 at https://pr35283-5b03ac7.ngbuilds.io/.

kapunahelewong commented on aio/content/guide/providers.md:
```diff
@@ -52,6 +52,14 @@ Any component created within a lazy loaded module’s context, such as by router
 
 Though you can provide services by lazy loading modules, not all services can be lazy loaded. For instance, some modules only work in the root module, such as the Router. The Router works with the global location object in the browser.
 
+With Angular version 9, you can provide a new instance of service with each lazy loaded module. The change to `UserService` would be as follows:
```
Oops! I must have missed this one:
```suggestion
With Angular version 9, you can provide a new instance of a service with each lazy loaded module. The change to `UserService` would be as follows:
```

kapunahelewong reviewed (commented):
It's looking great! One small comment and a question - why did you choose to move the `UserService` boxes inside the lazy loaded modules box? It looked good outside, like the one above it. If they're different, we may need to explain it/point it out (with only a sentence or two).

santoshyadav198613 commented:
> xplain it/point it out (with only a sentence or two)

I did it because the box is entire application, the image denotes that entire application uses same instance, and another image to show, that lazy loaded module is part of application and shares the unique instance per module.

mary-poppins commented:
You can preview c1c37ca at https://pr35283-c1c37ca.ngbuilds.io/.

mary-poppins commented:
You can preview 818c09d at https://pr35283-818c09d.ngbuilds.io/.

mary-poppins commented:
You can preview 4c056ca at https://pr35283-4c056ca.ngbuilds.io/.

santoshyadav198613 commented:
Hi @kapunahelewong ,
Can you review please.

santoshyadav198613 commented:
Hi @kapunahelewong ,
Can you review please.

santoshyadav198613 commented:
Hi Team,
Can you review it, please?

aikidave commented on aio/content/guide/providers.md:
```diff
@@ -52,6 +52,14 @@ Any component created within a lazy loaded module’s context, such as by router
 
 Though you can provide services by lazy loading modules, not all services can be lazy loaded. For instance, some modules only work in the root module, such as the Router. The Router works with the global location object in the browser.
 
+With Angular version 9, you can provide a new instance of a service with each lazy loaded module. The change to `UserService` would be as follows:
```
```suggestion
With Angular version 9, you can provide a new instance of a service with each lazy loaded module. The following code adds this functionality to `UserService`.
```

aikidave commented on aio/content/guide/providers.md:
```diff
@@ -52,6 +52,14 @@ Any component created within a lazy loaded module’s context, such as by router
+
+<code-example path="providers/src/app/user.service.2.ts"  header="src/app/user.service.ts"></code-example>
+
+With `providedIn: 'any'`, all eagerly loaded modules share a singleton instance, however lazy loaded modules each get their own unique instance, as in the following diagram.
```
```suggestion
With `providedIn: 'any'`, all eagerly loaded modules share a singleton instance; however, lazy loaded modules each get their own unique instance, as shown in the following diagram.
```

santoshyadav198613 commented:
Thanks for review @aikidave , all changes are done.

mary-poppins commented:
You can preview 3cd1d32 at https://pr35283-3cd1d32.ngbuilds.io/.

santoshyadav198613 commented:
Hi @aikidave , can you check now? let me know any changes are required.

santoshyadav198613 commented:
Hi, @alxhub @IgorMinar can you review, please.

kapunahelewong reviewed (approved):
Just realized I forgot to _submit_ my teeny review! Thank you, @santoshyadav198613!

Reviewed-for: global-docs-approvers

santoshyadav198613 commented:
Hi @kapunahelewong ,
Thanks for review done with the changes.

mary-poppins commented:
You can preview 0b3e206 at https://pr35283-0b3e206.ngbuilds.io/.

IgorMinar reviewed (approved):
LGTM. Thanks!

Reviewed-for: code-ownership

IgorMinar reviewed (commented):
btw @santoshyadav198613 I recommend that for the diagram we put each of the lazy loaded chunks into each own blue box. that way it doesn't look like they are all loaded together, and the diagram will be easier to digest.

santoshyadav198613 commented:
> btw @santoshyadav198613 I recommend that for the diagram we put each of the lazy loaded chunks into each own blue box. that way it doesn't look like they are all loaded together, and the diagram will be easier to digest.

Sure will make the changes.

santoshyadav198613 commented:
@IgorMinar please check now.

mary-poppins commented:
You can preview 33f89e9 at https://pr35283-33f89e9.ngbuilds.io/.

IgorMinar commented:
Hmm. Looks the same to me. Can you check?

santoshyadav198613 commented:
> Hmm. Looks the same to me. Can you check?

Oops my bad, misunderstood your reply. I will make the changes.

mary-poppins commented:
You can preview b775245 at https://pr35283-b775245.ngbuilds.io/.

santoshyadav198613 commented:
Hi @IgorMinar let me know if there is any changes needed

santoshyadavdev commented:
Hi Team,
Let me know if any changes are needed.

fmalcher commented:
Thank you @santoshyadavdev !
I still have som very minor concerns about the image:
- it's SVG but has no vector paths in it but just embedded PNG
- it's higher than wide so it could (!) look strange on the docs page in full width. IMO we could make it a bit wider so it looks a bit more asthetic
- some of the labels in the boxes are not aligned to the center

Which tool did you use to draw them? Let me know if I can help. I can redraw the image as SVG if you like. 🙂

santoshyadavdev commented:
Thank @fmalcher, yes I am very bad at creating images 😅 I used https://vectr.com/ to create this one. Would be great help if you can help with correct SVG 😊

fmalcher commented:
https://www.dropbox.com/s/og8e1p5qebrt47k/any-provider.svg?dl=1

What do you think? @santoshyadavdev @kapunahelewong

santoshyadavdev commented:
> ![Screen Shot 2020-07-08 at 19 52 12](https://user-images.githubusercontent.com/1683147/86953322-8fab4000-c154-11ea-92dd-9c3e0f53928b.png)
> 
> https://www.dropbox.com/s/og8e1p5qebrt47k/any-provider.svg?dl=1
> 
> What do you think? @santoshyadavdev @kapunahelewong

Much better than what I had :)

fmalcher commented:
great! use this minified SVG then, please 😅 https://www.dropbox.com/s/kfn6o2wbaxii0im/any-provider.svg?dl=0

santoshyadavdev commented:
> great! use this minified SVG then, please 😅 https://www.dropbox.com/s/kfn6o2wbaxii0im/any-provider.svg?dl=1
> @santoshyadavdev

Thanks @fmalcher Just a question I am getting a watermark in the image saying **Viewer does not support full svg 1.1**
![image](https://user-images.githubusercontent.com/11923975/86954002-e9226780-c172-11ea-9056-a6b4ae9c6178.png)

fmalcher commented:
> > great! use this minified SVG then, please 😅 https://www.dropbox.com/s/kfn6o2wbaxii0im/any-provider.svg?dl=1
> > @santoshyadavdev
> 
> Thanks @fmalcher Just a question I am getting a watermark in the image saying **Viewer does not support full svg 1.1** its not there in original image though
> ![image](https://user-images.githubusercontent.com/11923975/86954002-e9226780-c172-11ea-9056-a6b4ae9c6178.png)

Oops, got it... use this version then – and let me know if it doesn't work yet!
https://www.dropbox.com/s/og8e1p5qebrt47k/any-provider_.svg?dl=1

santoshyadavdev commented:
> > > great! use this minified SVG then, please 😅 https://www.dropbox.com/s/kfn6o2wbaxii0im/any-provider.svg?dl=1
> > > @santoshyadavdev
> > 
> > 
> > Thanks @fmalcher Just a question I am getting a watermark in the image saying **Viewer does not support full svg 1.1** its not there in original image though
> > ![image](https://user-images.githubusercontent.com/11923975/86954002-e9226780-c172-11ea-9056-a6b4ae9c6178.png)
> 
> Oops, got it... use this version then – and let me know if it still doesn't work!
> https://www.dropbox.com/s/og8e1p5qebrt47k/any-provider_.svg?dl=1

Yes, this worked thanks :)

santoshyadavdev committed:
docs: add explanation for providedIn any

Angular 9 introduces a new value for providedIn called `any` which lets us use unique instance
for servicec in each lazy loaded module, this PR is to document the same

fixes #35179

santoshyadavdev commented:
Thanks a lot, @fmalcher Pushed with the new image 👍

mary-poppins commented:
You can preview 8f14a74 at https://pr35283-8f14a74.ngbuilds.io/.

IgorMinar reviewed (approved):
looks good now. thanks for updating the diagram @santoshyadavdev and thanks for helping out @fmalcher ❤️
